### PR TITLE
fix: isolate empty-mention waiter by sender without changing default session scope

### LINKED ISSUE #9377: [Bug] 群聊中空提及等待器会截获其他成员的消息（session_waiter 未绑定发送者）
### What happened / 发生了什么

群聊中，成员 A 发送仅包含 `@机器人` 的空提及后，AstrBot 会启动 60 秒的
`session_waiter` 等待补充内容。但该等待器只以 `unified_msg_origin`（群会话）作为
会话标识，不包含发送者 ID。

因此在等待窗口内，同群**任意其他成员** B 发送的第一条非空消息会被等待器截获：

1. B 的原始消息被 `event.stop_event()` 终止传播（其他插件也收不到这条消息）；
2. AstrBot 在消息段开头人为插入指向机器人的 `At` 组件；
3. 浅复制后作为新事件重新投递，走完整流水线并触发 LLM 回复。

结果：一条完全没有提及机器人的普通群消息被当作 A 的补充输入，机器人向错误的
用户、以错误的上下文作出回复。

**根因定位**（已对照 master fb02c72 确认）：

- `astrbot/core/utils/session_waiter.py` 中 `DefaultSessionFilter.filter()` 仅返回
  `event.unified_msg_origin`，群聊中该值对所有成员相同；
- `astrbot/builtin_stars/astrbot/main.py` 的 `empty_mention_waiter` 未传入
  `session_filter`，使用了上述默认值。

另外，仅含唤醒前缀的消息（`is_wake_prefix_only` 分支）会启动同一个等待器，
跨用户截获问题同样适用于该场景。

**与文档矛盾**：`docs/zh/dev/star/guides/session-control.md`（"自定义会话 ID 算子"
一节）写明「默认情况下，AstrBot 会话控制器会将基于 `sender_id`（发送人的 ID）作为
识别不同会话的标识，如果想将一整个群作为一个会话，则需要自定义会话 ID 算子」——
即文档承诺默认按发送人隔离，实现却是按群隔离，两者恰好相反。

**回归引入点（git 考古）**：按发送人隔离是该功能的原始设计，当前行为是一次
后续改动引入的回归：

- `24c20a19` / `0caff054`（2025-03）：会话控制诞生时 `DefaultSessionFilter`
  即为 `return event.get_sender_id()`，docstring 写明"返回发送者的 ID"，
  文档与之匹配；
- `4aa91ad5`（2025-03-08）：空提及功能的提交信息原文为"支持当消息只有@bot时，
  **下一条发送人的消息**直接唤醒机器人"——功能规格本身即按发送人；
- `cfae6550`（2025-04-19，"perf: 修改默认会话过滤器标识符为umo"）：两行改动
  将 key 从 `get_sender_id()` 换为 `unified_msg_origin`，文档与空提及调用点
  均未随之调整，跨用户截获自此引入。

需要说明：`cfae6550`（对应 [#1326](https://github.com/AstrBotDevs/AstrBot/pull/1326)，
动机原文为「避免默认标识符导致的跨会话问题」）修复的是 v3 key 的一个真实
缺陷——仅含 sender_id 的 key 会让同一用户在 A 群创建的等待器截获其在
B 群/私聊的消息（跨会话串线）。该 PR 当时未讨论群聊内多用户共享 key 的
副作用。因此**简单 revert 会复活旧 bug**；正确的 key 需要同时包含会话与
发送者两部分，下述两条修复路线均满足这一点。

**连带问题**（同一处代码，修复时建议一并考虑）：

- `handle_session_control_agent` 中 `event.stop_event()` 无条件执行，而等待器对
  `message_str` 为空的消息（纯图片/语音/文件/表情）直接 `return` 且不结束会话。
  因此等待窗口内任意成员发送的纯图片都会被静默吞掉，且等待器继续存活，
  60 秒内可反复发生（实测日志见下方"补充实验"段）；
- `SessionWaiter._cleanup()` 无条件 `USER_SESSIONS.pop(self.session_id)`，同 key 的
  旧等待器超时清理时会把新注册的等待器一并移除。

### Reproduce / 如何复现？

前提：`platform_settings.unique_session` 保持默认关闭（开启后 WakingCheckStage
会将发送者并入 session_id，本问题会被掩蔽——这也解释了为何部分部署观察不到）。

1. 群聊中启用空提及等待（`platform_settings.empty_mention_waiting`，默认开启）；
2. 成员 A 发送仅含 `@机器人` 的消息（无正文）；
3. 在 60 秒等待窗口内（注意：窗口从机器人回复完成后开始计时），
   同群另一成员 B 发送任意带文字的普通消息（不 @ 机器人）；
4. 观察：B 的消息被添加机器人 At 后重新投递，触发 LLM 回复；B 的原始消息
   不再正常传播。

补充：第 3 步若 B 发送的是纯图片，则该消息被静默吞掉且等待器不结束。

### AstrBot version, deployment method (e.g., Windows Docker Desktop deployment), provider used, and messaging platform used. / AstrBot 版本、部署方式（如 Windows Docker Desktop 部署）、使用的提供商、使用的消息平台适配器

- AstrBot v4.25.5（已确认最新 v4.26.7 与 master fb02c72 中相关逻辑未变化）
- 部署方式：Windows Docker Desktop 部署（Docker Desktop	4.74.0 (227015) Docker Engine	29.4.3，API 1.54）
- 提供商：OpenAI 兼容接口
- 消息平台：aiocqhttp（NapCat）

### OS

Windows

### Logs / 报错日志

以下为一次受控复现的完整日志窗口（两个测试账号：`hedssaz/3315540459` 为成员 A，
`1/49983336` 为成员 B）。由于根因已定位至源码且可按上述步骤稳定复现，此处为
问题时间窗内的完整 INFO 级日志；如需 Debug 级日志可随时补充。

**对照组 —— 同一用户补充内容（设计内的正常路径）：**

```text
[2026-07-24 20:09:43.625] event_bus: hedssaz/3315540459: [At:3878611411]      ← A 发空 @
[2026-07-24 20:09:47.674] respond.stage: Prepare to send - hedssaz/3315540459: [引用消息] 喵？这么晚还没睡呀~ ...   ← 60 秒窗口自此开始
[2026-07-24 20:10:45.696] event_bus: hedssaz/3315540459: 1                    ← A 自己补充“1”（无 At）
[2026-07-24 20:10:45.704] respond.stage: Prepare to send - hedssaz/3315540459:（空）
[2026-07-24 20:10:45.704] context_utils: astrbot - after_message_sent 终止了事件传播。   ← 原始事件被吞
[2026-07-24 20:10:45.705] event_bus: hedssaz/3315540459: [At:3878611411] 1    ← 内部插入 At 后重投
[2026-07-24 20:10:49.184] respond.stage: Prepare to send - hedssaz/3315540459: [引用消息] 喵？收到一个「1」！...
```

**实验组 —— 另一用户的消息被截获（bug 本体）：**

```text
[2026-07-24 20:10:57.240] event_bus: hedssaz/3315540459: [At:3878611411]      ← A 发空 @
[2026-07-24 20:11:02.085] respond.stage: Prepare to send - hedssaz/3315540459: [引用消息] 铃在这里呀~ ...   ← 窗口开始
[2026-07-24 20:11:05.896] event_bus: 1/49983336: 1                            ← B 发“1”，无任何 At 消息段
[2026-07-24 20:11:05.905] respond.stage: Prepare to send - 1/49983336:（空）
[2026-07-24 20:11:05.905] context_utils: astrbot - after_message_sent 终止了事件传播。   ← B 的原始事件被吞
[2026-07-24 20:11:05.905] event_bus: 1/49983336: [At:3878611411] 1            ← B 的消息被插入 At 后重投
[2026-07-24 20:11:05.906] respond.stage: Prepare to send - hedssaz/3315540459:（空）
[2026-07-24 20:11:05.906] context_utils: astrbot - after_message_sent 终止了事件传播。   ← A 的空 @ 事件随等待器结束收尾
[2026-07-24 20:11:10.249] respond.stage: Prepare to send - 1/49983336: [引用消息] 又是「1」喵！你们是在排队点铃吗~
```

日志中可见的三个关键事实：

1. B 的消息进入事件总线时（20:11:05.896）不含任何 At 段，9ms 后（.905）即以
   `[At:3878611411] 1` 的形态重新出现——该 At 不可能来自平台往返，只能是
   AstrBot 内部插入；
2. 20:11:05.905/.906 两行空发送分别属于两个不同用户：B 的原始消息被 A 创建的
   等待器消费并终止，A 的空 @ 事件同时收尾；
3. 机器人最终回复中的“你们”表明 A 与 B 的输入已被混入同一对话上下文。

两组唯一的变量是补充消息的发送者，行为完全一致——证明等待器不区分发送者。

**补充实验 —— 连带问题"消息吞没"的实测（纯图片被静默丢弃）：**

```text
[2026-07-24 20:34:43.020] event_bus: 1/49983336: [At:3878611411]              ← B 发空 @，起等待器
[2026-07-24 20:34:48.365] respond.stage: Prepare to send - 1/49983336: [引用消息] 喵？铃在呢~ ...   ← 窗口开始
[2026-07-24 20:34:59.210] event_bus: hedssaz/3315540459: [图片]               ← A 发纯图片（无文字）
[2026-07-24 20:34:59.219] respond.stage: Prepare to send - hedssaz/3315540459:（空）
[2026-07-24 20:34:59.219] context_utils: astrbot - after_message_sent 终止了事件传播。   ← 图片被吞，机器人无任何回应
[2026-07-24 20:35:19.484] event_bus: hedssaz/3315540459: 111                  ← 20 秒后 A 发文字
[2026-07-24 20:35:19.491] respond.stage: Prepare to send - hedssaz/3315540459:（空）
[2026-07-24 20:35:19.491] context_utils: astrbot - after_message_sent 终止了事件传播。   ← 原始文字事件被吞
[2026-07-24 20:35:19.491] event_bus: hedssaz/3315540459: [At:3878611411] 111  ← 插入 At 后重投
[2026-07-24 20:35:19.492] respond.stage: Prepare to send - 1/49983336:（空）
[2026-07-24 20:35:19.492] context_utils: astrbot - after_message_sent 终止了事件传播。   ← B 的空 @ 事件随等待器结束收尾
[2026-07-24 20:35:23.060] respond.stage: Prepare to send - hedssaz/3315540459: [引用消息] 三个「1」！...
```

这一段同时证实了两点：

1. **纯图片被静默吞没**：A 的 `[图片]`（20:34:59.210）在 9ms 后出现
   空发送 + 终止传播的组合（.219），机器人无任何回应；若图片是正常传播，
   不会产生这两行。且等待器未被图片终结——20 秒后 A 的文字仍被捕获，
   说明窗口期内每条非文本消息都会被反复丢弃；
2. **吞没同样不区分发送者**：本段中起等待器的是 B（1/49983336），被吞掉
   图片和被截获文字的却是 A（hedssaz）——与主 bug 同源。

（最初在真实群聊中发现本问题时的表现与上述一致：真实成员的普通消息
“应该吧”被插入 At 并触发回复，NapCat 侧原始事件确认 atType=0、无 At 消息段。）

### Are you willing to submit a PR? / 你愿意提交 PR 吗？

- [x] Yes!

### Code of Conduct

- [x] I have read and agree to abide by the project's [Code of Conduct](https://docs.github.com/zh/site-policy/github-terms/github-community-code-of-conduct)。

### PR DESCRIPTION
**Pull request by hedssaz**
Fixes #9377

In group chats, an empty mention (`@bot` with no text) starts a 60s
`session_waiter` to wait for follow-up content. The waiter is registered with
`DefaultSessionFilter`, whose key is `unified_msg_origin` (group-level) and does
not include the sender. As a result, the first non-empty message from **any
other member** during the wait window is intercepted, gets a bot `At` prepended,
and is re-dispatched — producing a reply to the wrong user / wrong context. Full
analysis, reproduction and captured logs are in #9377.

This PR fixes the empty-mention flow **without changing the default session
scope**, so third-party plugins that rely on the current group-level default are
unaffected.

### Why not simply change `DefaultSessionFilter`

Changing the default from `unified_msg_origin` to include the sender would
change the default runtime behavior of `session_waiter`, which is exported to
plugins through `astrbot.api.util`. The group-level default has been in place
since v3.5.4 (~15 months, 138 version tags) and is relied upon by real
plugins that omit `session_filter=`, for example:

- `astrbot_plugin_kimi_datasource_api` registers a waiter manually with
  `SessionWaiter(DefaultSessionFilter(), event.unified_msg_origin, ...)`. The
  registration key is the raw UMO while lookups go through
  `DefaultSessionFilter.filter(event)`. If the default changed, the keys would no
  longer match and its login waiter would **never trigger**.
- `astrbot_plugin_decision_roulette` and `astrbot_plugin_TurtleSoup` use the
  default filter to collect input from multiple group members; a per-sender
  default would only accept the initiator.

So the fix is scoped to the one place that actually needs per-sender isolation.

### Modifications

- `astrbot/core/utils/session_waiter.py`: add `SenderSessionFilter`, a new
  optional `SessionFilter` keyed by `unified_msg_origin` + `sender_id`.
  `DefaultSessionFilter` is **unchanged**, and no new names are added to the
  public `astrbot.api.util` surface. This is purely additive.
  The key uses a length-prefixed encoding (`f"{len(umo)}:{umo}:{sender}"`)
  rather than plain concatenation: both the UMO and the sender id are arbitrary
  platform strings (e.g. Telegram topic-group UMOs contain `#`), so a plain
  separator would allow constructible key collisions that defeat the isolation.
- `astrbot/builtin_stars/astrbot/main.py`: the empty-mention waiter now passes
  `session_filter=SenderSessionFilter()`. When the sender id is unavailable it
  skips waiting entirely, to avoid silently degrading back to group scope.
- `tests/test_session_waiter_cross_user.py` (new): 7 session-key regression
  tests (module loaded standalone).
- `tests/unit/test_empty_mention_sender_scope.py` (new): 3 production-wiring
  tests driving the real `Main.handle_empty_mention` and
  `Main.handle_session_control_agent` (following the existing
  `Main.__new__(Main)` pattern in `tests/unit/test_group_chat_context_wiring.py`).

The other issues noted in #9377 (empty-`message_str` messages being swallowed
during the wait window; `_cleanup` / `FILTERS` behavior with same-key waiters)
are independent of the session key and will be handled in separate PRs.

### Screenshots or Test Results

```
$ ruff format --check <changed files + test files> && ruff check <same>
All checks passed!

$ pytest tests/test_session_waiter_cross_user.py \
         tests/unit/test_empty_mention_sender_scope.py -o asyncio_mode=auto -q
..........                                                               [100%]
10 passed
```

Session-key coverage (`tests/test_session_waiter_cross_user.py`):

1. `test_default_filter_returns_umo_unchanged` — the default filter still returns
   `unified_msg_origin` (locks the default semantics in place);
2. `test_default_manual_umo_registration_still_triggers` — reproduces the
   `kimi_datasource_api` manual-registration pattern and asserts it still
   triggers (guards against regressing plugins that depend on the group default);
3. `test_sender_filter_isolates_users_in_group` — another member's message is no
   longer intercepted (primary goal);
4. `test_sender_filter_accepts_original_sender` — the original sender's follow-up
   is still captured;
5. `test_sender_filter_key_is_collision_free` — UMOs / sender ids containing
   separator characters (`#`, `:`) cannot produce colliding keys;
6. `test_sender_filter_isolates_users_with_hash_in_umo` — isolation holds for
   Telegram topic-group style UMOs that contain `#`;
7. `test_sender_filter_does_not_leak_across_conversations` — the same user's
   messages in other groups / DMs are not intercepted.

Production-wiring coverage (`tests/unit/test_empty_mention_sender_scope.py`,
drives the real handlers rather than a re-implementation):

8. `test_other_members_message_passes_through` — BOB's message is not stopped,
   not modified, and not re-enqueued while ALICE's waiter is active;
9. `test_original_senders_followup_is_redispatched` — ALICE's follow-up gets the
   bot `At` injected, is stopped and re-enqueued, and the waiter is cleaned up;
10. `test_empty_sender_id_skips_waiting` — with no reliable sender id the
    handler finishes immediately and registers no waiter.

Verified that the wiring tests actually guard the fix: with the `main.py`
change reverted to master, tests 8 and 10 fail; with the fix in place all 10
pass.

---

- [x] This is NOT a breaking change.

> The default session scope is unchanged; only the built-in empty-mention waiter
> opts into per-sender isolation. `SenderSessionFilter` is an additive core
> utility and is not exported through `astrbot.api.util`.

### Checklist

- [x] 😊 If new features are added, they have been discussed with the authors
  via issues/emails. (Bug fix, proposed in #9377.)
- [x] 👀 My changes have been well-tested, with verification steps and results
  provided above.
- [x] 🤓 No new dependencies are introduced.
- [x] 😮 My changes do not introduce malicious code.

## Summary by Sourcery

Fix empty-mention session handling in group chats by isolating the waiter per sender while preserving existing group-level session semantics.

Bug Fixes:
- Prevent empty-mention waiters in group chats from intercepting follow-up messages from other members and replying in the wrong context.
- Ensure empty-mention handling skips session waiting when no reliable sender ID is available to avoid unintended group-scoped behavior.

Enhancements:
- Introduce an optional SenderSessionFilter that scopes session keys by conversation origin and sender using a collision-safe encoding without changing the default session filter.
- Strengthen session key behavior with explicit guarantees against key collisions and cross-conversation leakage.

Tests:
- Add regression tests that lock in DefaultSessionFilter behavior and validate SenderSessionFilter isolation, collision safety, and cross-conversation separation.
- Add production wiring tests for empty-mention handling to verify correct per-sender redispatch, non-interference with other members, and behavior when sender ID is missing.

### CODE REVIEW

**Review by xiaoyuyu6420 — COMMENTED** [review]
## Review

I've reviewed this PR in the context of issue #9377 alongside the two other open PRs (#9422, #9442) that address the same bug. Here's my analysis.

### Strengths

1. **Length-prefixed key encoding** (`{len(umo)}:{umo}:{sender_id}`) — this is technically the most robust approach to key collision resistance. Telegram topic-group origins contain `#`, and `:` appears in all UMO formats (`platform:MessageType:id`), so plain concatenation with a separator has a theoretical collision surface. This encoding is uniquely decodable.

2. **Comprehensive test suite** — two separate test files covering both unit-level key logic and integration-level `Main.handle_empty_mention` wiring. The regression guard for `astrbot_plugin_kimi_datasource_api` (manual UMO registration pattern) is particularly valuable.

3. **Opt-in design preserves backward compatibility** — `DefaultSessionFilter` is unchanged, so third-party plugins that rely on group-level keys are not affected.

### Issues

**1. Does not fix the non-text message swallowing bug (mentioned in #9377 as a "连带问题")**

The issue report explicitly documents that pure image messages during the wait window are silently dropped and the waiter stays alive:

```python
empty_mention_waiter handler:
    if not event.message_str or not event.message_str.strip():
        return  # ← pure image is silently dropped, waiter keeps running
```

This PR only adds `SenderSessionFilter` to the call site but does not modify the handler body. After this fix, Alice's pure image in the wait window is still silently consumed (by Alice herself now, not by Bob — an improvement, but the message is still lost).

**2. Does not fix the `_cleanup` race condition (mentioned in #9377 as a "连带问题")**

The issue report documents that `SessionWaiter._cleanup()` unconditionally calls `USER_SESSIONS.pop(self.session_id)`. If a newer waiter is registered under the same key before the old one times out, the old waiter's cleanup evicts the new one. This PR does not address this.

**3. Skipping the waiter on empty `sender_id` is a behavioral regression**

```python
if str(event.get_sender_id()):
    await empty_mention_waiter(event, session_filter=SenderSessionFilter())
```

When `sender_id` is empty/falsy, the waiter is simply not started. On platforms where `get_sender_id()` may temporarily return empty, the bot silently stops responding to empty mentions with no feedback to the user. A fallback placeholder (e.g., `<unknown>`) with a warning log would be safer — it degrades gracefully instead of silently dropping functionality.

**4. Only fixes `empty_mention_waiter` call site, not `is_wake_prefix_only`**

The issue report notes: *"仅含唤醒前缀的消息（is_wake_prefix_only 分支）会启动同一个等待器，跨用户截获问题同样适用于该场景"*. The same `session_waiter` is used in both branches, but this PR only passes `SenderSessionFilter` to the `empty_mention_waiter` call. The wake-prefix-only path still uses `DefaultSessionFilter` and is still vulnerable to cross-user interception.

### Summary

This PR correctly identifies the core isolation problem and proposes a clean opt-in API with excellent test coverage. However, it addresses only one of the three issues documented in #9377 (cross-user interception), leaving the non-text swallowing and cleanup race unfixed. The empty-sender-id skip also introduces a minor behavioral regression.

File: astrbot/core/utils/session_waiter.py [hunk 1]
```diff
@@ -101,6 +101,41 @@ def filter(self, event: AstrMessageEvent) -> str:
...
+        prefix makes the encoding uniquely decodable.
+
+        Args:
+            event: The current message event.
+
+        Returns:
+            An unambiguous session key of the form "{len(umo)}:{umo}:{sender_id}".
+        """
+        umo = event.unified_msg_origin
+        return f"{len(umo)}:{umo}:{event.get_sender_id()}"
```
[the remark this answers is not in the document]
Comment (hedssaz) [reply astrbot/core/utils/session_waiter.py]: Fixed in 12c1fc9 — sender keys are now prefixed with `!sender:`, which can never be parsed as a valid UMO since platform ids reject `!` (`PlatformManager._is_valid_platform_id`).

**Comment by hedssaz** [comment]
@xiaoyuyu6420 Thanks for the detailed review — and for cross-referencing the other PRs.

**Review points 1 and 2 (non-text swallowing, `_cleanup` race)** — correct, and intentional.

I reported all three problems in #9377 and split them deliberately so the session-key change can land without also changing semantics that plugins depend on. I'm happy to send follow-up PRs, or fold those fixes into this one if the maintainers prefer.

**Review point 4 (wake-prefix path)** — I believe this path is already covered.

There is exactly one `@session_waiter` in the file and one call site. Both branches merge at the shared guard:

```python
if not (is_empty_mention or is_wake_prefix_only):
    return
```

Everything after that guard follows the same path and reaches:

```python
await empty_mention_waiter(
    event,
    session_filter=SenderSessionFilter(),
)
```

There is no second call that still uses `DefaultSessionFilter`. Both paths are also subject to the same requirement that a usable sender ID be available. If you have a scenario where the wake-prefix path does not receive the sender-scoped filter, I'll add a regression test.

**Review point 3 (empty `sender_id`)** — agreed, and thanks for catching the missing log; **added**, with a test.

Two clarifications:

* With the default configuration (`empty_mention_waiting_need_reply` enabled), the bot still replies. That reply is emitted before the waiter is registered, so only the subsequent 60-second wait is skipped, not the response itself.
* On the `<unknown>` placeholder: two members who both yield an empty sender ID would collide on `{umo}:<unknown>`, reintroducing cross-user interception for exactly the users whose identities cannot be verified. Skipping fails closed (a feature is unavailable); the placeholder fails open (the wrong user's message may be consumed). For a bug whose symptom is replying to the wrong person, failing closed seems safer.

## Why this PR leaves `DefaultSessionFilter` alone

To be upfront, changing the default probably would not break many plugins. For most of them, per-sender keying would be neutral or better — “reply with the song number,” “enter the verification code,” and “reply `确认` to proceed” are all flows where only the initiator should answer, whereas today any member of the group can do so.

I did find plugins whose behavior would change.

`astrbot_plugin_decision_roulette` collects options through the default waiter, so today any member who replies in the group can add one; with per-sender keying, only the initiator's replies would be accepted.

`astrbot_plugin_kimi_datasource_api` constructs its waiter as:

```python
SessionWaiter(
    DefaultSessionFilter(),
    pending.session_id,
    False,
)
```

Here, `pending.session_id` is the raw UMO. Registration therefore uses that raw key, while lookups go through `DefaultSessionFilter.filter(event)`. Changing the default would make the two keys permanently differ, causing its chat-control waiter — which handles `cancel`, `restart`, and `status` during login — to stop receiving messages.

Its separate OAuth polling task could still complete the login, so this would be a partial regression rather than a total failure.

What concerns me is not the number of affected plugins, but the fact that this failure is silent: no error is raised; the waiter simply stops firing. Candidate call sites can be found through static search, but whether a given plugin requires group scope or sender scope is encoded in its workflow logic rather than its call signature, so the complete affected set cannot be determined statically.

That feels like a change to make deliberately, with a changelog entry, rather than implicitly inside a bug fix. Either way, it is a maintainer decision. This PR keeps the blast radius small while leaving that decision open.

For what it's worth, the current documentation describes a per-sender default, while the implementation has returned `unified_msg_origin` since #1326 / `cfae6550`. I can only date the earliest public version of that documentation page to 2025-11-16, so I cannot determine which behavior came first.

#9422 also keeps `DefaultSessionFilter` unchanged and scopes the fix to the built-in empty-mention waiter.

**Summary:** Review point 4 is already covered; the warning and test requested in review point 3 have been added; and review points 1 and 2 are accurate but intentionally scoped out. Whichever approach is chosen for #9377, I would suggest retaining a regression test for the manual UMO-registration pattern, since that failure mode is silent.

**Comment by xiaoyuyu6420** [comment]
@hedssaz Thanks for the thorough response — and for the correction on point 4.

**You're right about point 4.** I re-read the control flow: both `is_empty_mention` and `is_wake_prefix_only` merge after the guard at line 72 and reach the single `empty_mention_waiter(event)` call at line 143. There is no second call site using `DefaultSessionFilter`. My comment was factually wrong — I apologize for the noise.

**On point 3 (`<unknown>` vs skip):** You make a strong argument. The distinction between "fails closed" (feature unavailable) and "fails open" (wrong user's message consumed) is the right framing. For a bug whose symptom is replying to the wrong person, failing closed is indeed safer. I retract that suggestion — your approach is better for this specific bug class.

The only scenario where the placeholder would be preferable is if `get_sender_id()` returning empty is common on some platform (making the feature frequently unavailable). But that's speculative without data, and the conservative choice is correct.

**On points 1 and 2 (intentional scope split):** Makes sense. Keeping the session-key change independent of the semantic changes (non-text handling, cleanup race) lets each land on its own merit. If the maintainers prefer to address all three together, #9442 covers them; if they prefer the incremental approach, your PR + follow-ups is cleaner.

**On the regression test for manual UMO registration:** Strongly agree. Whichever approach wins, a test guarding against the silent failure mode (registration key ≠ lookup key) should be retained. I've added a similar guard in #9442's `TestDecoratorPathCompatibility`, but yours in `test_default_manual_umo_registration_still_triggers` is more direct because it tests the actual breakage scenario.

Appreciate the constructive discussion — this is exactly the kind of review dialogue that makes the end result better.

**Comment by xiaoyuyu6420** [comment]
Thanks for the thorough impact analysis — this is exactly the homework I could not do from the call sites alone. The `kimi_datasource_api` case is the decisive one for me: raw-UMO registration with a filtered lookup is precisely the pattern that would regress silently, and "partial regression during login flows" is the kind of bug nobody can diagnose from logs. That settles my concern about keeping the default untouched: changing it should be a deliberate, changelogged decision for maintainers, not a side effect of this fix.

Points 1 and 2 being intentionally scoped out is reasonable given #9377 already tracks them. +1 to retaining a regression test for the manual UMO-registration pattern whichever route is chosen — silent failure modes need a permanent test guard.

One follow-up worth splitting out: the docs-vs-implementation mismatch on the per-sender default (docs say per-sender, implementation has returned `unified_msg_origin` since #1326). That is an independent doc bug and would get lost inside this PR — worth its own issue so it stays visible after this merges.
